includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,384 @@
     }
     $filePath = drupal_realpath($archival_file['file']->uri);
     $failureMessage = array(
       'success' => false,
-      'messages' => array(
-        'message' => t('Failed to upload object @pid to Panopto, but no exception caught.'),
+      'messages' => array(),
+    );
+    $streaming = array();
+    
+    $oauthendpoint = variable_get('islandora_video_streaming_panopto_host');
+    $uploadendpoint = $oauthendpoint . '/Panopto/PublicAPI/Rest/sessionUpload';
+    $oauthendpoint .= "/Panopto/oauth2/connect/token";
+    $token = islandora_video_streaming_panopto_get_token($oauthendpoint);
+    $session = null;
+    if ($token) {
+      $session = islandora_video_streaming_panopto_get_session($token, $uploadendpoint); 
+    }
+    $success = false;
+    if ($session) {
+      $upload_id = $session['ID'];
+      $upload_target = $session['UploadTarget'];
+      $manifest_file = islandora_video_streaming_panopto_create_manifest_for_video($filePath);
+      if (isset($manifest_file)) {
+        try {
+          islandora_video_streaming_panopto_upload_file($upload_target, $manifest_file);
+          islandora_video_streaming_panopto_upload_file($upload_target, $filePath);
+        } catch (Exception $ex) {
+          $messages = array(
+            'message' => t('Panopto upload error: @exception'),
+            'message_sub' => array('@exception' => $ex->getMessage()),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+          watchdog_exception('islandora_video_streaming_panopto', $ex);
+        }
+        $response = islandora_video_streaming_panopto_finish_upload($uploadendpoint, $session, $token);
+        if (isset($response)) {
+          try {
+            $success = islandora_video_streaming_monitor_progress($uploadendpoint, $upload_id, $token);
+          } catch (Exception $ex) {
+            $messages = array(
+            'message' => t('Panopto progress error: @exception'),
+            'message_sub' => array('@exception' => $ex->getMessage()),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+            );
+            $failureMessage['messages'][] = $messages;
+            watchdog_exception('islandora_video_streaming_panopto', $ex);
+            return $failureMessage;
+          }
+          $streaming['upload_id'] = $session['ID'];
+        } else {
+          $messages = array(
+            'message' => t('Issue with finishing upload for object: @pid, response: @response'),
+            'message_sub' => array(
+              '@pid' => $object->id,
+              '@response' => $response,
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+        }
+      } 
+    } else {
+      $messages = array(
+        'message' => t('Failed to retrieve session for object: @pid'),
```

Comment:
This error message gives the user no real understanding of why the session failed.  Possibilities include:
* the `$token` evaluated to false.
  * but `islandora_video_streaming_panopto_get_token()` does not report details on why a `$token` is not returned.
  * (consider: `isset($oauthendpoint) && isset($client_id) && isset($secret_client_id) && isset($username) && isset($password)`)
* the call to `islandora_video_streaming_panopto_get_session()` failed
  * but `islandora_video_streaming_panopto_get_session()` does not itself provide full coverage for reporting possible errors.
  * (consider: `isset($folder_id)`)

An excellent option would be to have the functions `islandora_video_streaming_panopto_get_token()`  `islandora_video_streaming_panopto_get_session()` throw appropriate errors, which would then be handled in the caller.  A good option (involving less re-coding) would be to have functions `islandora_video_streaming_panopto_get_token()`  `islandora_video_streaming_panopto_get_session()` log specific errors to the watchdog, which then would correlate with the general error from the caller.